group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,903 @@
+        // Find the correct partition movement considering the stickiness requirement.
+        TopicIdPartition partitionToBeMoved = partitionMovements.getTheActualPartitionToBeMoved(
+            partition,
+            member,
+            newMember
+        );
+        processPartitionMovement(partitionToBeMoved, newMember);
+    }
+
+    private void processPartitionMovement(TopicIdPartition topicIdPartition, String newMember) {
+        String oldMember = partitionOwnerInTargetAssignment.get(topicIdPartition);
+
+        partitionMovements.movePartition(topicIdPartition, oldMember, newMember);
+
+        assignmentManager.removePartitionFromTargetAssignment(topicIdPartition, oldMember);
+        assignmentManager.addPartitionToTargetAssignment(topicIdPartition, newMember);
+    }
+
+    /**
+     * <code> MemberPair </code> represents a pair of member Ids involved in a partition reassignment. Each
+     * <code> MemberPair </code> object, which contains a source (<code>src</code>) and a destination (<code>dst</code>)
+     * element, normally corresponds to a particular partition or topic, and indicates that the particular partition or some
+     * partition of the particular topic was moved from the source member to the destination member during the rebalance.
+     */
+    private static class MemberPair {
+        private final String srcMemberId;
+        private final String dstMemberId;
+
+        MemberPair(String srcMemberId, String dstMemberId) {
+            this.srcMemberId = srcMemberId;
+            this.dstMemberId = dstMemberId;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((this.srcMemberId == null) ? 0 : this.srcMemberId.hashCode());
+            result = prime * result + ((this.dstMemberId == null) ? 0 : this.dstMemberId.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null)
+                return false;
+
+            if (!getClass().isInstance(obj))
+                return false;
+
+            MemberPair otherPair = (MemberPair) obj;
+            return this.srcMemberId.equals(otherPair.srcMemberId) && this.dstMemberId.equals(otherPair.dstMemberId);
+        }
+
+        @Override
+        public String toString() {
+            return "MemberPair(" +
+                "srcMemberId='" + srcMemberId + '\'' +
+                ", dstMemberId='" + dstMemberId + '\'' +
+                ')';
+        }
+    }
+
+    /**
+     * This class maintains some data structures to simplify lookup of partition movements among members.
+     * During a partition rebalance, it keeps track of partition movements corresponding to each topic,
+     * and also possible movement (in form a <code>MemberPair</code> object) for each partition.
+     */
+    private static class PartitionMovements {
```

Comment:
i'm still not sure why it's necessary to track the previous movements.

can you help me understand why we need this? Is this to prevent cyclic reassignments?

Reply:
yes that's precisely it, just to make sure we're not moving partitions back and forth mainly.